File: Python/0022-m-generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        list = []
        self.backtrack(list,"",0,0,n)
        return list
    
    def backtrack(self,list,str,open,close,max):
        if len(str) == max * 2:
            list.append(str)
            return
        if open < max:
            self.backtrack(list,str+"(",open+1,close,max)
        if close < open:
            self.backtrack(list,str+")",open,close+1,max)

File: leetcode/C++/0022-m-generate-parentheses.cpp
// Backtracking
class Solution {
private:
    void helper(int n, int open, int close, string& path, vector<string>& res) {
        if (open == n && close == n) {
            res.push_back(path);
            return;
        }
        if (open < n) {
            path += "(";
            helper(n, open + 1, close, path, res);
            path.pop_back();
        }
            
        if (close < open) {
            path += ")";
            helper(n, open, close + 1, path, res);
            path.pop_back();
        }
    }
    
public:
    vector<string> generateParenthesis(int n) {
        string path = "";
        vector<string> res;
        helper(n, 0, 0, path, res);
        return res;
    }
};